File: ObjCSample/Tests/ViewControllerPresentationTests.m
//  ViewControllerPresentationSpy by Jon Reid, https://qualitycoding.org
//  Copyright 2021 Quality Coding, Inc. See LICENSE.txt

#import "CodeNextViewController.h"

#import "StoryboardNextViewController.h"
#import "ViewController.h"

@import ViewControllerPresentationSpy;
@import XCTest;

@interface ViewControllerPresentationTests : XCTestCase
@end

@implementation ViewControllerPresentationTests

{
    QCOPresentationVerifier *presentationVerifier;
    ViewController *sut;
}

- (void)setUp
{
    [super setUp];
    presentationVerifier = [[QCOPresentationVerifier alloc] init];
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    sut = [storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([ViewController class])];
    [sut loadViewIfNeeded];
}

- (void)tearDown
{
    [NSRunLoop.currentRunLoop runUntilDate:NSDate.date]; // Free objects after segue show
    presentationVerifier = nil;
    sut = nil;
    [super tearDown];
}

- (void)test_outlets_shouldBeConnected
{
    XCTAssertNotNil(sut.seguePresentModalButton);
    XCTAssertNotNil(sut.segueShowButton);
    XCTAssertNotNil(sut.codePresentModalButton);
}

- (void)test_tappingSeguePresentModalButton_shouldPresentNextViewControllerWithGreenBackground
{
    [sut.seguePresentModalButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqual(presentationVerifier.presentedCount, 1, @"presented count");
    XCTAssertEqual(presentationVerifier.presentingViewController, sut, @"presenting view controller");
    XCTAssertTrue(presentationVerifier.animated, @"animated");
    if (![presentationVerifier.presentedViewController isKindOfClass:[StoryboardNextViewController class]])
    {
        XCTFail(@"Expected presented view controller to be %@, but was %@",
                [StoryboardNextViewController class], presentationVerifier.presentedViewController);
        return;
    }
    StoryboardNextViewController *nextVC = (StoryboardNextViewController *)presentationVerifier.presentedViewController;
    XCTAssertEqual(nextVC.backgroundColor, UIColor.greenColor, @"Background color passed in");
}

- (void)test_tappingSegueShowButton_shouldShowNextViewControllerWithRedBackground
{
    UIWindow *window = [[UIWindow alloc] init];
    window.rootViewController = sut;
    window.hidden = NO;

    [sut.segueShowButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqual(presentationVerifier.presentedCount, 1, @"presented count");
    XCTAssertEqual(presentationVerifier.presentingViewController, sut, @"presenting view controller");
    XCTAssertTrue(presentationVerifier.animated, @"animated");
    if (![presentationVerifier.presentedViewController isKindOfClass:[StoryboardNextViewController class]])
    {
        XCTFail(@"Expected presented view controller to be %@, but was %@",
                [StoryboardNextViewController class], presentationVerifier.presentedViewController);
        return;
    }
    StoryboardNextViewController *nextVC = (StoryboardNextViewController *)presentationVerifier.presentedViewController;
    XCTAssertEqual(nextVC.backgroundColor, UIColor.redColor, @"Background color passed in");
}

- (void)test_tappingCodeModalButton_shouldPresentNextViewControllerWithPurpleBackground
{
    [sut.codePresentModalButton sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqual(presentationVerifier.presentedCount, 1, @"presented count");
    XCTAssertEqual(presentationVerifier.presentingViewController, sut, @"presenting view controller");
    XCTAssertTrue(presentationVerifier.animated, @"animated");
    if (![presentationVerifier.presentedViewController isKindOfClass:[CodeNextViewController class]])
    {
        XCTFail(@"Expected presented view controller to be %@, but was %@",
                [ViewController class], presentationVerifier.presentedViewController);
        return;
    }
    CodeNextViewController *nextVC = (CodeNextViewController *)presentationVerifier.presentedViewController;
    XCTAssertEqual(nextVC.backgroundColor, UIColor.purpleColor, @"Background color passed in");
}

@end

File: SampleSwift/Tests/ViewControllerPresentationTests.swift
@testable import MockUIAlertControllerSampleSwift
import XCTest

final class ViewControllerPresentationTests: XCTestCase {
    private var presentationVerifier: QCOMockPresentationVerifier!
    private var sut: ViewController!

    override func setUp() {
        super.setUp()
        presentationVerifier = QCOMockPresentationVerifier()
        let storyboard = UIStoryboard.init(name: "Main", bundle: nil)
        sut = storyboard.instantiateInitialViewController() as? ViewController
        sut.loadViewIfNeeded()
    }

    override func tearDown() {
        presentationVerifier = nil
        sut = nil
        super.tearDown()
    }

    func test_outlets_shouldBeConnected() {
        XCTAssertNotNil(sut.codePresentModalButton)
        XCTAssertNotNil(sut.seguePresentModalButton)
    }

    func test_tappingSeguePresentModalButton_shouldPresentNextViewControllerWithGreenBackground() {
        sut.seguePresentModalButton.sendActions(for: .touchUpInside)

        XCTAssertEqual(presentationVerifier.presentedCount, 1, "presented count")
        XCTAssertTrue(presentationVerifier.animated, "animated")
        XCTAssertTrue(presentationVerifier.presentingViewController === sut,
                """
                Expected presenting view controller to be \(String(describing: sut)), \
                but was \(presentationVerifier.presentingViewController)
                """)
        guard let nextVC = presentationVerifier.presentedViewController as? StoryboardNextViewController else {
            XCTFail("""
                    Expected presented view controller to be \(String(describing: StoryboardNextViewController.self)), \
                    but was \(presentationVerifier.presentedViewController)
                    """)
            return
        }
        XCTAssertEqual(nextVC.backgroundColor, UIColor.green, "Background color passed in")
    }

    func test_tappingCodeModalButton_shouldPresentNextViewControllerWithPurpleBackground() {
        sut.codePresentModalButton.sendActions(for: .touchUpInside)

        XCTAssertEqual(presentationVerifier.presentedCount, 1, "presented count")
        XCTAssertTrue(presentationVerifier.animated, "animated")
        XCTAssertTrue(presentationVerifier.presentingViewController === sut,
                """
                Expected presenting view controller to be \(String(describing: sut)), \
                but was \(presentationVerifier.presentingViewController)
                """)
        guard let nextVC = presentationVerifier.presentedViewController as? CodeNextViewController else {
            XCTFail("""
                    Expected presented view controller to be \(String(describing: CodeNextViewController.self)), \
                    but was \(presentationVerifier.presentedViewController)
                    """)
            return
        }
        XCTAssertEqual(nextVC.backgroundColor, UIColor.purple, "Background color passed in")
    }
}
